New CCE utilities needed for CCE-GH executable

moxcodes opened the pull request:
## Proposed changes

Add extra interface manager tweaks and communication actions to allow for simultaneous evolution of globally time-stepped Generalized Harmonic with a differently time-stepped CCE evolution.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies

- [x] #2289

moxcodes committed:
Add interp strategies to Cce::InterfaceManagers

moxcodes committed:
Fix entry in worldtube component

moxcodes committed:
Separate interface insert next time

moxcodes committed:
Add action for sending next time to Cce

moxcodes committed:
Remove mistaken include

moxcodes committed:
fix clang-tidy

fmahebert commented on src/Evolution/Systems/Cce/InterfaceManagers/GhLocalTimeStepping.cpp:
```diff
@@ -16,6 +16,9 @@
 #include "Time/TimeStepId.hpp"
+#include "Utilities/BoostHelpers.hpp"
+
+#include "Parallel/Printf.hpp"
```
Is this left over from debugging?

fmahebert commented on src/Evolution/Systems/Cce/InterfaceManagers/GhLocalTimeStepping.cpp:
```diff
@@ -65,8 +83,54 @@ void GhLocalTimeStepping::insert_gh_data(
+      update_history();
+    }
+    return;
+  }
```
[aesthetics nitpick] I would find this easier to parse if you avoided the double-nested conditionals, perhaps like,
```
if (previous_deque_entry == pre_history_.end()) {
  // work
  return;
} else if (the other conditions) {
  // other work
  return;
}
```
or am I missing something about the logic that favors the current structure? Anyway, highly nitpicky, up to you what you prefer here 🙂

fmahebert commented on tests/Unit/Evolution/Systems/Cce/InterfaceManagers/Test_GhLocalTimeStepping.cpp:
```diff
@@ -121,8 +120,19 @@ void test_gh_local_time_stepping_interface_manager(
+              local_interface_manager,
+          const size_t index) noexcept {
+        auto current_time_step_id = source_time_steps[index];
+        auto next_time_step_id = source_time_steps[index + 1];
```
could/should these be const?

fmahebert commented on src/Evolution/Systems/Cce/Actions/SendNextTimeToCce.hpp:
```diff
@@ -0,0 +1,152 @@
+      }
+    }
+    if (first_valid_block_logical_coords.empty() ) {
+      Parallel::printf(
```
Should this file include Parallel/Printf.hpp instead?

fmahebert commented on src/Evolution/Systems/Cce/Actions/SendNextTimeToCce.hpp:
```diff
@@ -0,0 +1,152 @@
+      Parallel::simple_action<Actions::ReceiveNextElementTime>(
+          receiver_proxy, db::get<::Tags::TimeStepId>(box),
+          std::move(next_step_id));
+      return std::forward_as_tuple(std::move(box));
```
Is this line needed? Would omitting it not go to the same statement on line 100?

fmahebert commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_SendNextTimeToCce.cpp:
```diff
@@ -0,0 +1,205 @@
+          substep},
+      std::make_unique<::TimeSteppers::RungeKutta3>());
+
+  // check that the 'next' time made it into the interpolation manager
```
Is this an extra comment, or a missing check below the comment?

moxcodes replied:
oops, yep, this is an extra comment before I decided to put all the checks into the previous function.

fmahebert reviewed (commented):
Some small observations...

moxcodes commented:
fixups posted.
Cheers!

geoffrey4444 commented on src/Evolution/Systems/Cce/Initialize/InverseCubic.hpp:
```diff
@@ -3,8 +3,6 @@
 
 #pragma once
 
```
Just curious... why was this include there? it seems circular to me

moxcodes replied:
I think this must have just been a copy-paste mistake, it is certainly circular.

geoffrey4444 commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -319,11 +328,34 @@ struct EndTime : db::SimpleTag {
+
+  static constexpr bool pass_metavariables = false;
+  static double create_from_options() {
+    return std::numeric_limits<double>::infinity();
```
just curious...can this ever lead to an FPE? if not, why not?

moxcodes replied:
In current code, this won't FPE because it is only used in the comparison operator to see if the current time is still less than the end time. It would if we tried to do math with it in the future.

geoffrey4444 commented on src/Evolution/Systems/Cce/Actions/SendNextTimeToCce.hpp:
```diff
@@ -0,0 +1,152 @@
+            db::get<intrp::Tags::InterpPointInfo<Metavariables>>(box));
+    const std::vector<ElementId<Metavariables::volume_dim>> element_ids{
+        {array_index}};
+    typename intrp::Vars::PointInfoTag<InterpolationTargetTag,
```
Is this equivalent to `using first_valid_block_logical_coords = ...`? If so is one way preferred?

moxcodes replied:
This one is constructing an object of the type represented by the `PointInfoTag<...>::type`. I could use `db::item_type<>` instead, but I think we've been discouraging `item_type` where we can be confident it's not a compute tag.

geoffrey4444 commented on src/Evolution/Systems/Cce/Actions/SendNextTimeToCce.hpp:
```diff
@@ -0,0 +1,152 @@
+        element_ids, first_valid_block_logical_coords);
+
+    if (LIKELY(element_coord_holders.count(element_ids[0]) == 0)) {
+      return std::forward_as_tuple(std::move(box));
```
Why not combine the ifs to avoid duplicating the return statement?

geoffrey4444 commented on src/Evolution/Systems/Cce/InterfaceManagers/GhInterpolationStrategies.hpp:
```diff
@@ -0,0 +1,35 @@
+  if(strategy == InterpolationStrategy::EverySubstep) {
+    return true;
+  }
+  if(strategy == InterpolationStrategy::EveryStep) {
```
Design question: why don't strategies have should_interpolate() member functions, so this could be generic? Now if a new strategy is made, you have to remember to update here too

moxcodes replied:
This is a little bit difficult, because the interface manager is specified at runtime (by options), so whether we want every step vs every substep (or something else) is also only known at runtime, so the `InterpolationStrategy` has to be a runtime quantity. And, it has to be available to the DG element array, where the interface manager is stored on the cce worldtube component. The solution here was to just have the enum `InterpolationStrategy` in the const global cache.

However, now that you ask, a similarly effective solution would be to make the `InterpolationStrategy` a factory-creatable abstract base class, and each individual version a derived class, then the `std::unique_ptr` can sit in the const global cache, and have a `should_interpolate` function. Do you think that would better?

moxcodes replied:
oops, I'm wrong -- this is less direct than I thought because it would need the `should_interpolate` function to be a virtual, but it wouldn't be able to be a template like it is here.

geoffrey4444 replied:
It sounds like it’s a hassle, so let’s leave as is. Thanks for the explanation!

geoffrey4444 commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_GhBoundaryCommunication.cpp:
```diff
@@ -175,9 +175,10 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.Cce.Actions.GhBoundaryCommunication",
+          l_max, extraction_radius, end_time, start_time,
+          InterfaceManagers::InterpolationStrategy::EveryStep,
+          number_of_radial_points,
+          std::make_unique<::TimeSteppers::RungeKutta3>()}};
```
Why RK3? AB1 is faster, and for real evolutions I usually use DP5

moxcodes replied:
I think I just chose this to make sure things work with substeps, but it could just as easily be DP5.

moxcodes replied:
Swapped to DP5 in the dependency PR #2289.

geoffrey4444 reviewed (changes requested):
Just a few questions

moxcodes commented:
rebased on develop, no longer dependent.
Cheers!

moxcodes commented:
@geoffrey4444 @fmahebert shall I squash?

fmahebert commented:
Ok from me! 👍

geoffrey4444 commented:
Please squash

moxcodes commented:
Fixups squashed in.
Cheers!

kidder commented on src/Evolution/Systems/Cce/InterfaceManagers/GhLocalTimeStepping.cpp:
```diff
@@ -45,17 +46,32 @@ void GhLocalTimeStepping::insert_gh_data(
 
+  // retrieve an iterator position if the next time has already been
+  // inserted for this data
+  const auto previous_deque_entry = std::find_if(
```
use `alg::find_if`

kidder commented on src/Evolution/Systems/Cce/InterfaceManagers/GhLocalTimeStepping.cpp:
```diff
@@ -65,8 +81,54 @@ void GhLocalTimeStepping::insert_gh_data(
+    TimeStepId time_id, TimeStepId next_time_id) noexcept {
+  // retrieve an iterator position if the next time has already been
+  // inserted for this data
+  const auto previous_deque_entry = std::find_if(
```
`alg::find_if`

kidder commented on tests/Unit/Evolution/Systems/Cce/InterfaceManagers/Test_GhLocalTimeStepping.cpp:
```diff
@@ -121,8 +120,19 @@ void test_gh_local_time_stepping_interface_manager(
+          const size_t index) noexcept {
+        const auto current_time_step_id = source_time_steps[index];
+        const auto next_time_step_id = source_time_steps[index + 1];
+        local_interface_manager->insert_next_gh_time(current_time_step_id,
```
just inline the above two statements into the `insert_next_gh_time` call

kidder commented on src/Evolution/Systems/Cce/Actions/SendNextTimeToCce.hpp:
```diff
@@ -0,0 +1,150 @@
+
+namespace Cce::Actions {
+
+struct ReceiveNextElementTime;
```
cond this

kidder commented on src/Evolution/Systems/Cce/Actions/SendNextTimeToCce.hpp:
```diff
@@ -0,0 +1,150 @@
+    const std::vector<ElementId<Metavariables::volume_dim>> element_ids{
+        {array_index}};
+    typename intrp::Vars::PointInfoTag<InterpolationTargetTag,
+                                       Metavariables::volume_dim>::type
```
do you know the actual type?  looks like it should just be a container of Tensor...

kidder commented on src/Evolution/Systems/Cce/Actions/SendNextTimeToCce.hpp:
```diff
@@ -0,0 +1,150 @@
+struct ReceiveNextElementTime {
+  template <typename ParallelComponent, typename... DbTags, typename ArrayIndex,
+            typename Metavariables,
+            Requires<tmpl::list_contains_v<tmpl::list<DbTags...>,
```
remove this and use a `if constexpr` where the false branch includes a useful ERROR message

kidder commented on tests/Unit/Evolution/Systems/Cce/Actions/CMakeLists.txt:
```diff
@@ -14,6 +14,7 @@ set(LIBRARY_SOURCES
   Test_InitializeCharacteristicEvolution.cpp
   Test_InitializeWorldtubeBoundary.cpp
   Test_RequestBoundaryData.cpp
+  Test_SendNextTimeToCce.cpp
```
alphabet

kidder commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp:
```diff
@@ -26,6 +26,7 @@
 #include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp"
 #include "Utilities/FileSystem.hpp"
 #include "Utilities/Gsl.hpp"
+#include "Domain/ElementId.hpp"
```
alphabet

moxcodes replied:
I squashed the fix for this one in because the line in question changed somewhere else in the commit history.

kidder commented on src/Evolution/Systems/Cce/InterfaceManagers/GhInterpolationStrategies.hpp:
```diff
@@ -0,0 +1,35 @@
+namespace Cce {
+namespace InterfaceManagers{
+
+enum class InterpolationStrategy { EveryStep, EverySubstep };
```
document this

kidder commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -378,6 +378,19 @@ struct GhInterfaceManager : db::SimpleTag {
+
+  static constexpr bool pass_metavariables = false;
+  static InterfaceManagers::InterpolationStrategy
+  create_from_options(
```
We are confused about who choose the strategy...  It looks like it should be the InterfaceManager, but here you are allowing it to be option-creatable ...

moxcodes replied:
Yeah, I can see how this logic is a little confusing.
The reason I put it together this way is that the interpolation strategy needs to be in the const global cache, but it is determined by which interface manager is chosen by the options, and the interface manager does not end up in the const global cache because it stores stateful data that should be held only by the worldtube component. 

So my solution was to have this option creatable, but reference the same option tag used to make the interface manager, so that it can get the appropriate interpolation strategy when the const global cache is initialized from options.
But, I admit that this is a bit confusing, so I'm happy to implement any alternatives that seem less tangled.

kidder reviewed (changes requested):
reviewed with @nilsdeppe

moxcodes commented:
I've pushed fixups for all of the comments except for the option-creatable interpolation strategy question.
Cheers!

moxcodes committed:
Change alg::find functions to forward container

moxcodes commented:
fixed the build problems, and added a commit (`Change alg::find functions to forward container`) that changes the Utilities/Algorithm.hpp to std::forward the potentially non-const references now, because in my usage of the `std::find`, I wanted to modify the value after the I located the desired entry in the deque.

kidder commented on src/Evolution/Systems/Cce/InterfaceManagers/GhInterpolationStrategies.hpp:
```diff
@@ -9,8 +9,26 @@
+ *
+ * \details The available strategies are:
+ * - `InterpolationStrategy::EveryStep` : interpolation data will be provided on
+ * each full step for the time-stepper, so the worldtube data represents
```
could you please indent each line of the items by two more space (so it is aligned with the `)

kidder reviewed (commented):
Please add a comment about the option_creatable tag.
You can squash.

moxcodes commented:
Squashed the fixups along with the comment changes.
Cheers!
